### ISSUE TITLE
error compiling wasm func ... failed to compile: BUG: relative jump offset 

### ISSUE THREAD

**Issue by ydnar** [tetratelabs/wazero#1522]
**Describe the bug**
Ran a program with the most recent tagged version of `wazero`, which failed with this error:

```sh
error: error compiling wasm func[.github.com_zonedb_zonedb.init]: failed to compile: BUG: relative jump offset -2097240/4(=-524310) must be within -524288 and 524287
```

https://github.com/tetratelabs/wazero/blob/b01ba67fdcf32997d22a0f56b723a71c1c40cd75/internal/asm/arm64/impl.go#L864

Prior to upgrading `wazero`, the program ran correctly.

**To Reproduce**
I was compiling a program that imports https://github.com/zonedb/zonedb.

**Expected behavior**
For the program to compile and run.

**Environment (please complete the relevant information):**
- Go version: `gotip` as of 2023-06-15
- wazero Version: v1.2.1
- Host architecture: arm64
- Runtime mode: compiler

**Additional context**
Using `wasirun` from https://github.com/stealthrocket/wasi-go

**Comment by mathetake** [comment]
https://github.com/tetratelabs/wazero/pull/1516 I think this is the root cause since this only happens in v1.2.1 cc @evacchi @ncruces

**Comment by mathetake** [comment]
thanks for the report @ydnar, let me clarify that this hasn't happened before v1.2.1, correct?

**Comment by ydnar** [comment]
I confirmed the error occurs with `wazero` @ f3858732 (#1516), but not in the immediate parent commit 26eae5756.

**Comment by mathetake** [comment]
cool, thank you for confirmation!

**Comment by ydnar** [comment]
> cool, thank you for confirmation!

No, thank you for writing this! Between y’all and @stealthrocket, I’ve been able to port a relatively large network service to `wasip1` in a few days.

**Comment by evacchi** [comment]
are you able to share a reproducer @ydnar ?

I wonder if a large function may trigger that error since we are now jumping to the first trap handler.

**Comment by ncruces** [comment]
Yes, that must be it: a function over 2Mb.

**Comment by ncruces** [comment]
This file fails the same: 
[huge.wasm.gz](https://github.com/tetratelabs/wazero/files/11768894/huge.wasm.gz)

On `amd64`:
```sh
./wazero run huge.wasm
OK
```

On `arm64`:
```sh
./wazero run huge.wasm                                                        
error compiling wasm binary: error compiling wasm func[.__original_main]: failed to compile: BUG: relative jump offset -2531896/4(=-632974) must be within -524288 and 524287
```

The WASM comes using `wasi-sdk` to build this (except it has 8k repetitions of those lines):
```c
#include <stdio.h>

volatile int counter;

int main(void) {
        counter *= 5;
        counter++;
        counter *= 5;
        counter++;
        counter *= 5;
        counter++;
        …
        counter *= 5; 
        counter++; 
        puts("OK\n"); 
        return counter != 0; 
}
```

I can't really test a fix, because I don't have a `arm64`, I simply hacked wazero to spit `arm64` it then can't run.

**Comment by ncruces** [comment]
The immediate fix is:
```go
func (c *arm64Compiler) compileMaybeExitFromNativeCode(skipCondition asm.Instruction, status nativeCallStatusCode) {
	skip := c.assembler.CompileJump(skipCondition)
	c.compileExitFromNativeCode(status)
	c.assembler.SetJumpTargetOnNext(skip)
}
```

Can we test the performance of it compared to `1.2.0` and `1.2.1`?

**Comment by ncruces** [comment]
Right. That's the immediate fix, if you feel that it's important to roll out a patch. That, or revert #1516.

**Comment by evacchi** [comment]
interesting, the patch causes `TestCompiler_compileBrTable` to fail

```
--- FAIL: TestCompiler_compileBrTable (0.00s)
    --- FAIL: TestCompiler_compileBrTable/only_default_with_index_0 (0.00s)
        require.go:330: expected 1, but was 6
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:316
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:477
    --- FAIL: TestCompiler_compileBrTable/only_default_with_index_100 (0.00s)
        require.go:330: expected 1, but was 6
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:316
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:477
    --- FAIL: TestCompiler_compileBrTable/select_default_with_targets_and_good_index (0.00s)
        require.go:330: expected 1, but was 4
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:316
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:477
    --- FAIL: TestCompiler_compileBrTable/select_default_with_targets_and_huge_index (0.00s)
        require.go:330: expected 1, but was 4
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:316
            /Users/evacchi/Devel/github.com/tetratelabs/wazero/internal/engine/compiler/compiler_controlflow_test.go:477
```

**Comment by ncruces** [comment]
That's weird (the test failure, and your fix).

Isn't that test (`target := c.compiledTrapTargets[status]; target != nil`) exactly what's at the top of `compileExitFromNativeCode`? How is that different?

**Comment by ncruces** [comment]
Also it occurs to me that we're missing a `return` in these two places:
https://github.com/tetratelabs/wazero/blob/b01ba67fdcf32997d22a0f56b723a71c1c40cd75/internal/engine/compiler/impl_arm64.go#L417-L419
https://github.com/tetratelabs/wazero/blob/b01ba67fdcf32997d22a0f56b723a71c1c40cd75/internal/engine/compiler/impl_amd64.go#L4857-L4859

Surely there's no point in emitting the rest of the code if we're doing an unconditional jump before it. Making this change may change the perf numbers (less code, better prediction, yes even _unconditional_ jumps involve branch prediction [these days](https://blog.cloudflare.com/branch-predictor/)).

**Comment by ydnar** [comment]
Thanks, everyone. I appreciate all the attention you've given this!

**Comment by evacchi** [comment]
I didn't realize that they were exactly the same 🤔

**Comment by evacchi** [comment]
> That's weird (the test failure, and your fix).

I most likely made a mistake.

**Comment by ncruces** [comment]
3% is nothing to sneeze at, if we can't do better.
At least it's a fix?

**Comment by evacchi** [comment]
sure, it's something :P I'd add a proper test too so we won't get caught into this issue in the future (I think we should be able to generate a wasm function directly 🤔)
